packages/yew-router-macro/src/routable_derive.rs

```diff
@@ -0,0 +1,217 @@
+use proc_macro2::{Span, TokenStream};
+use quote::quote;
+use syn::parse::{Parse, ParseStream};
+use syn::punctuated::Punctuated;
+use syn::spanned::Spanned;
+use syn::{Data, DeriveInput, Fields, Ident, LitStr, Variant, Visibility};
+
+const AT_ATTR_IDENT: &str = "at";
+const NOT_FOUND_ATTR_IDENT: &str = "not_found";
+
+pub struct Routable {
+    vis: Visibility,
+    ident: Ident,
+    ats: Vec<LitStr>,
+    variants: Punctuated<Variant, syn::token::Comma>,
+    not_found_route: Option<LitStr>,
+}
+
+impl Parse for Routable {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let DeriveInput {
+            vis, ident, data, ..
+        } = input.parse()?;
+
+        let data = match data {
+            Data::Enum(data) => data,
+            Data::Struct(s) => {
+                return Err(syn::Error::new_spanned(
+                    s.struct_token,
+                    "expected enum, found struct",
+                ))
+            }
+            Data::Union(u) => {
+                return Err(syn::Error::new_spanned(
+                    u.union_token,
+                    "expected enum, found union",
+                ))
+            }
+        };
+
+        let (not_found_route, ats) = parse_variants(&data.variants)?;
+
+        Ok(Self {
+            vis,
+            ident,
+            variants: data.variants,
+            ats,
+            not_found_route,
+        })
+    }
+}
+
+fn parse_variants(
+    variants: &Punctuated<Variant, syn::token::Comma>,
+) -> syn::Result<(Option<LitStr>, Vec<LitStr>)> {
+    let mut not_founds = vec![];
+    let mut ats: Vec<LitStr> = vec![];
+
+    for variant in variants.iter() {
+        let variant: &syn::Variant = variant;
+
+        if let Fields::Unnamed(_) = variant.fields {
+            return Err(syn::Error::new(
+                variant.span(),
+                "only named fields are supported",
+            ));
+        }
+
+        let attrs = &variant.attrs;
+        let at_attrs = attrs
+            .iter()
+            .filter(|attr| attr.path.is_ident(AT_ATTR_IDENT))
+            .collect::<Vec<_>>();
+
+        let attr = match at_attrs.len() {
+            1 => *at_attrs.first().expect("fucked"),
```

Comment:
😛 might want to replace this at some point...

Reply:
Oh boi... I was using it to debug and forgot to remove it. I removed it now.